node/src/main/java/io/airlift/node/NodeInfo.java

```diff
@@ -300,14 +307,59 @@ private static String findInternalAddress(AddressSource addressSource)
         };
     }
 
-    private static InetAddress findInternalIp()
+    private InetAddress findInternalIp()
+    {
+        List<Function<List<InetAddress>, Optional<InetAddress>>> searchOrder = new ArrayList<>(3);
+        if (this.preferIpv6Address) {
+            searchOrder.add(NodeInfo::findIpv6address);
+            searchOrder.add(NodeInfo::findLocalAddress);
+            searchOrder.add(NodeInfo::findIpv4address);
+        }
+        else {
+            searchOrder.add(NodeInfo::findLocalAddress);
+            searchOrder.add(NodeInfo::findIpv4address);
+            searchOrder.add(NodeInfo::findIpv6address);
+        }
+
+        List<InetAddress> goodAddresses = getGoodAddresses();
+        return searchOrder.stream()
+            .map(source -> source.apply(goodAddresses))
+            .filter(Optional::isPresent)                        // select only valid results
+            .findFirst()
+            .orElse(Optional.empty())                     // we could not find any source
+            .orElse(null);                                // it is most likely that this is a disconnected developer machine
+    }
+
+    private static Optional<InetAddress> findIpv4address(List<InetAddress> candidates)
+    {
+        // check all up network interfaces for a good v4 address
+        for (InetAddress address : candidates) {
+            if (isV4Address(address)) {
+                return Optional.of(address);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<InetAddress> findIpv6address(List<InetAddress> candidates)
+    {
+        // check all up network interfaces for a good v6 address
+        for (InetAddress address : candidates) {
+            if (isV6Address(address)) {
+                return Optional.of(address);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<InetAddress> findLocalAddress(List<InetAddress> candidates)
```

Comment:
@jalpan-randeri couple of comments/questions:
- Should this be called `findLocalIpv4Address`? It only looks for IPv4, right?
- `candidates` is not used in this function
- Why do we only look for an IPv4 local addr? Why not IPv6 as well (if `preferIpv6Address == true`)?